Add `ktlint` Kotlin linter

FloEdelmann opened the pull request:
*Before reading on: Yes, there are an awful lot of changes in this PR. But I think it's worth it.*

### About ktlint

[ktlint](https://github.com/pinterest/ktlint) is an easy-to-use linter for Kotlin. Most of the reported issues can also be fixed automatically. It tries to implement the [official Kotlin code style](https://kotlinlang.org/docs/coding-conventions.html) without having to configure anything.

It is configurable though: Rules we don't like [can be disabled](https://github.com/pinterest/ktlint#custom-ktlint-specific-editorconfig-properties) through `.editorconfig`.

And it is extensible: We could write a [custom ruleset](https://github.com/pinterest/ktlint#creating-a-ruleset) someday that e.g. checks the property order in quest type classes (see #3526 and https://github.com/streetcomplete/StreetComplete/commit/2fcfd101cbe01c84f6e9b7549c0ed887e015c6e2) – with a real AST, not with hacky regexes.

### How to run ktlint

* The ktlint binary can be installed and run manually as a CLI tool. That is how I did it locally.
	* Then just run `ktlint` in the repo root to check all Kotlin files, or `ktlint --format` to fix most reported issues automatically.
	* You can run `ktlint --apply-to-idea-project` to adjust the formatting settings of Android Studio to match what `ktlint` would produce.
	* You can run `ktlint --install-git-pre-commit-hook` or `ktlint --install-git-pre-push-hook` to automatically run `ktlint` before `git commit`ing / `git push`ing, so lint issues don't find their way into the repo at all.
* There are [Maven and Gradle integrations](https://github.com/pinterest/ktlint#integration). I haven't checked those, as I'm not that familiar with those tools.
* On GitHub Actions, we can run ktlint on certain events with the [ktlint-check](https://github.com/musichin/ktlint-check) action. I added it to this PR.
	* Please see [my demo PR](https://github.com/FloEdelmann/StreetComplete/pull/508) for how that looks: The check fails because there are lint issues; a GitHub comment with an explanation is created automatically in this case. See [here](https://github.com/FloEdelmann/StreetComplete/runs/4855642710?check_suite_focus=true) for the workflow output. Go to the [files tab](https://github.com/FloEdelmann/StreetComplete/pull/508/files) to see the lint issues inline as annotations directly in the code changes. Also, there are a few unchanged files with check annotations at the bottom – that shouldn't happen if lint issues are regularly fixed/prevented.
	* The workflow is configured to run in every PR and on every push to the `master` branch. That way, lint issues should be caught quite quickly.

### Why we should add it

A more consistent codebase is more maintainable for everyone, because there are less decisions to think about. People that are new to the project learn the code style faster, because they are not confused by existing code that actually violates it. There is less to check in code reviews, because the lint issues are caught by an automatic tool already. So it might mitigate https://github.com/streetcomplete/StreetComplete/pull/3573#issuecomment-1005060213 a bit.

### What the linter complains about

That is what makes this PR so huge: There are many parts in the codebase that violate the ktlint code style. I have let ktlint fix most of the issues automatically, and I have not reviewed the changes thoroughly. I did notice some ugly formatting though, e.g. [here](https://github.com/FloEdelmann/StreetComplete/blob/96dea182a2eba0a65329128e44e69723324d91fc/app/src/main/java/de/westnordost/streetcomplete/quests/leaf_detail/AddForestLeafType.kt#L14-L24). I'll fix that later, if needed.

### How to proceed

1. Question to @westnordost and others: Do you like the idea of an automated linter at all?
2. If yes, then we need to decide if we want to disable any rules. I'd say no, but let's hear your opinions. To help with the decision, please go through this PR's commits that start with `Lint:` and see what code changes the respective rule causes. Please also see [this comment](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015692558) for discussions about specific rules.
3. The rules that we keep enabled should be fixed, and the resulting formatting should look good. That means, that I myself will go through the commits once more, and check the output thoroughly. Where needed, I'll add more manual lint changes that are in line with the code style.
4. We need to decide what to do with open PRs. The big list of changes will cause merge conflicts all over the place. Should we take the time once to fix the conflicts in all open PRs? I'd volunteer to do that where I have the privileges to do so.

matkoniecz commented:
In principle I like linters. I have some preferences but do not care about them too much. And I really prefer to have automatically formatted code over dealing with it manually. And focus on more important things.

And standard code formatting reveals some bugs.

-----------------

My main worry is that experience for someone making PR, especially new, should not become worse. 

I am not fan of situation where opening issue or PR is Kafkaesque scenario requires dealing with cryptic linter bot, CLA bot and things gets closed anyway with "inactive for 10 minutes, closing as stale". This is obviously not so bad but starts in this direction.

Someone will see

![screen06](https://user-images.githubusercontent.com/899988/149916315-169973d2-45ce-41b0-98d2-002fe3a88fdb.png)

Is it possible to make this to not look like something terrible happened? Or provide more detail how to run linter autoformatter?

On clicking details:

> Error: Error: ktlint exited with code 1

On expanding "ktlint check":

`  app/src/main/java/de/westnordost/streetcomplete/quests/leaf_detail/AddForestLeafType.kt:45:51: Unexpected spacing before ":" (colon-spacing)`
`Error: Unexpected spacing before ":"` 

Which is the expected spacing?

Is it possible to have bot opening PR against PR with linting applied? Or to make inline code suggestions?

Or at least to point out how this check can be run?

Additional issue is that it increases initial setup even more if user needs to install one more thing.

Maybe autoformatting should be mentioned and trigger warning, not a failure state on a PR? Or apply it automatically with github actions? Or have a script that we could run and would push commits into PR branches with linting? (I prefer to not give maintainer access to a bot account).

> There are Maven and Gradle integrations. I haven't checked those, as I'm not that familiar with those tools.

Making it easier to setup/run would be nice.

-----------------

I am not entirely sure about some enabled rules, but I prefer to keep it for later if discussing at all.

FloEdelmann commented:
Thanks for your feedback. I've updated the workflows so that in PRs, they will always pass:
![grafik](https://user-images.githubusercontent.com/202916/149926018-bbe5093c-e068-49ef-a6a1-78d520f27a45.png)
(Unfortunately, we can't add inline information there, like if there are lint issues or not.)

In the files tab, the inline annotations are still being added. Please see the [demo PR](https://github.com/FloEdelmann/StreetComplete/pull/508/files).

matkoniecz commented:
> In the files tab, the inline annotations are still being added. Please see the demo PR.

I thought about proposed edits, so that user is not forced to decode linter errors. Or is it impossible?

> Unfortunately, we can't add inline information there, like if there are lint issues or not.

Is it possible to at least name it in way that points to documentation how autoformatter can be run?

> Thanks for your feedback. I've updated the workflows so that in PRs, they will always pass:

That in turn looks like linter found no problems :(

matkoniecz commented on .github/ktlint-message.md:
```diff
@@ -0,0 +1,6 @@
+This project follows the [official Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html). We use [ktlint](https://github.com/pinterest/ktlint) to check that, and unfortunately, there is some code that violates the code style. Please go to this PR's *Files* tab and have a look at the inline annotations in the code. To see a list of all lint issues, click *Details* near th ...
```
```suggestion
This project follows the [official Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html). We use [ktlint](https://github.com/pinterest/ktlint) to check that, and unfortunately, there is some code that mismatches the code style. Please go to this PR's *Files* tab and have a look at the inline annotations in the code. To see a list of all lint issues, click *Details* near the failing "❌ Lint / Kotlin" check below.
```

matkoniecz replied:
I think that less aggressive word would be better.

Maybe mention also

> just formatting the code inside Android Studio now fixes most lint issues

?

FloEdelmann replied:
Thanks, I applied the nicer wording.

Your other suggestion is already accounted for in the next sentence:

> You can fix most issues automatically by [reformatting the code in Android Studio](https://stackoverflow.com/q/40926920).

matkoniecz commented:
Ooo, thanks for that - it is much nicer.

matkoniecz commented:
https://github.com/streetcomplete/StreetComplete/pull/3649/commits/511be48811088be8fcdb496a385ec58879130151 - is it going to be undone by Android Studio?

Sometimes its "autofix missing import" is doing this.

Nitpicking ahead:

https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9f3b543c07d771712a2d78b872602dfb99898328#diff-a8807ccc4e1eb01614ea84f1d3ca5b4e2771a0bc9bb0bf5b80b6c2e324eaa703 - is it an improvement to remove alignment?

https://github.com/streetcomplete/StreetComplete/pull/3649/commits/8896b71ac51a6310dab029d57f3b0cb80e82a8a9#diff-448c3a7b38f3e45c44bfd79b501653da5f5b0a6669f4e55f2f1c0c403663b4a9 - `&&` prefixing is less annoying when you add one more rule - more obvious to add, nicer diffs

https://github.com/streetcomplete/StreetComplete/pull/3649/commits/5964e1467bd29770e861071661937c2c0bd289a2#diff-27bdddb4ab370a8bd4531b218643d24a3388efdf7879f8b7748faf17cebeab5d - is it an improvement to remove alignment?

--------------------------------
Here I probably only need to adjust to the new style, but it feels really weird for now. Though if it is some official Kotlin style I probably should really get over it and spend my weirdness budget on something else.

https://github.com/streetcomplete/StreetComplete/pull/3649/commits/3170b0967aed66858ab8bb69250dd03229cd3bf5

https://github.com/streetcomplete/StreetComplete/pull/3649/commits/8256af0e3c1e2c34d601c68d81222be8351b46b0#diff-d9d8c6059afa9b60f81813f0a92f3142e66fdb4aa52fbc2b6c8726113901d924 `when (` vs `when(` - somehow I really prefer `when(`

----------------------------------

Other are really, really great.

FloEdelmann commented:
> is it going to be undone by Android Studio? Sometimes its "autofix missing import" is doing this.

Actually, I fixed it manually by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>o</kbd> (optimize imports) in Android Studio. There were only a few cases where that didn't work; there I removed one wildcard import at a time and used Android Studio's import suggestions. So it seems that Android Studio won't reintroduce wildcard imports by itself.

> is it an improvement to remove alignment?

Hmmm. Both rules (`no-multi-spaces` and `paren-spacing`) catch a lot of tiny mistakes, but also remove deliberate formatting. However, such custom spacing is always fiddly: When adding a new line to such places, you often need to add spaces to all other lines to keep everything aligned. Sometimes, with long names/conditions, it is actually less readable. Also, I can never remember whether to put the spaces before or after an `->` arrow. See also https://github.com/streetcomplete/StreetComplete/pull/3582#discussion_r766667403. So I think it's best to actually forbid it (that's a clear rule at least). As I said, I will go over all places touched by this PR again and manually improve the formatting. I think we can often find a style that passes these rules and also keeps the readability.

> `&&` prefixing is less annoying when you add one more rule - more obvious to add, nicer diffs

Indeed. I think I'm also voting for disabling this rule.

> Here I probably only need to adjust to the new style

I actually already tried to conform to exactly these rules until now. As I learned the code style from looking at other code in the project, this must have been predominant already (though inconsistent as well). But I also think the rules makes sense: `when` is like `if`, so spacing should be the same. Colons for variable typing have a different meaning than colons before superclasses/interfaces, so spacing should differ.

> Other are really, really great.

Nice to hear! :blush:

FloEdelmann commented:
I'll summarize the discussions about specific rules in this comment, and link to the original comments from here. The they can be hidden/collapsed to keep this thread manageable.

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/21bb71189fcfe459198471337220173e7d72fc6d"><code>21bb711</code></a>: remove empty lines before closing curly brace (<code>no-blank-line-before-rbrace</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/5c3da208db9084d4b9ac63816d64abe06e23f111"><code>5c3da20</code></a>: Collapse consecutive empty lines (<code>no-consecutive-blank-lines</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/b62784ec2dca44d13c0aacff6ca926953d51d4e2"><code>b62784e</code></a>: Add newline to end of files (<code>final-newline</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/3170b0967aed66858ab8bb69250dd03229cd3bf5"><code>3170b09</code></a>: Standardize spacing around colons (<code>colon-spacing</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): Here I probably only need to adjust to the new style
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): I actually already tried to conform to exactly these rules until now. As I learned the code style from looking at other code in the project, this must have been predominant already (though inconsistent as well). But I also think the rules makes sense: `when` is like `if`, so spacing should be the same. Colons for variable typing have a different meaning than colons before superclasses/interfaces, so spacing should differ.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/7433d3342d907cbe3202afd2534fc24e249dd053"><code>7433d33</code></a>: Standardize spacing around commas (<code>comma-spacing</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/3ac056e00afc5ca66f9726479676662adc686798"><code>3ac056e</code></a>: Standardize spacing around curly braces (<code>curly-spacing</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/8256af0e3c1e2c34d601c68d81222be8351b46b0"><code>8256af0</code></a>: Standardize spacing around keywords (if, else, when, …) (<code>keyword-spacing</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): Here I probably only need to adjust to the new style
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): I actually already tried to conform to exactly these rules until now. As I learned the code style from looking at other code in the project, this must have been predominant already (though inconsistent as well). But I also think the rules makes sense: `when` is like `if`, so spacing should be the same. Colons for variable typing have a different meaning than colons before superclasses/interfaces, so spacing should differ.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/64fe90ce409f141f7f7652ea83b3ce1f81287e0b"><code>64fe90c</code></a>: Standardize spacing around operators (+, *, !=, …) (<code>op-spacing</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/5964e1467bd29770e861071661937c2c0bd289a2"><code>5964e14</code></a>: Standardize spacing around parentheses (<code>paren-spacing</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): Nitpicking ahead: is it an improvement to remove alignment?
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): Hmmm. Both rules (`no-multi-spaces` and `paren-spacing`) catch a lot of tiny mistakes, but also remove deliberate formatting. However, such custom spacing is always fiddly: When adding a new line to such places, you often need to add spaces to all other lines to keep everything aligned. Sometimes, with long names/conditions, it is actually less readable. Also, I can never remember whether to put the spaces before or after an `->` arrow. See also https://github.com/streetcomplete/StreetComplete/pull/3582#discussion_r766667403. So I think it's best to actually forbid it (that's a clear rule at least). As I said, I will go over all places touched by this PR again and manually improve the formatting. I think we can often find a style that passes these rules and also keeps the readability.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/e191c1fed9e7c74136398e6eb1e9e01ac2ec476d"><code>e191c1f</code></a>: Standardize spacing around comments (<code>comment-spacing</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/57eaaad396da4dfaaed9e71dd5bde161b2a567d7"><code>57eaaad</code></a>: Standardize parameter list wrapping (<code>parameter-list-wrapping</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/8896b71ac51a6310dab029d57f3b0cb80e82a8a9"><code>8896b71</code></a>: Standardize logic operator chaining position (<code>chain-wrapping</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): Nitpicking ahead: `&&` prefixing is less annoying when you add one more rule - more obvious to add, nicer diffs
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): Indeed. I think I'm also voting for disabling this rule.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9c2f170cfb7c138d01501e0c7ccbecc54f41b511"><code>9c2f170</code></a>: Standardize indentation (<code>indent</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9f3b543c07d771712a2d78b872602dfb99898328"><code>9f3b543</code></a>: Collapse multiple consecutive spaces (<code>no-multi-spaces</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): Nitpicking ahead: is it an improvement to remove alignment?
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): Hmmm. Both rules (`no-multi-spaces` and `paren-spacing`) catch a lot of tiny mistakes, but also remove deliberate formatting. However, such custom spacing is always fiddly: When adding a new line to such places, you often need to add spaces to all other lines to keep everything aligned. Sometimes, with long names/conditions, it is actually less readable. Also, I can never remember whether to put the spaces before or after an `->` arrow. See also https://github.com/streetcomplete/StreetComplete/pull/3582#discussion_r766667403. So I think it's best to actually forbid it (that's a clear rule at least). As I said, I will go over all places touched by this PR again and manually improve the formatting. I think we can often find a style that passes these rules and also keeps the readability.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/6b9db3b0ae33a75d45a3be8a0bb77f0440c67552"><code>6b9db3b</code></a>: Simplify variables in string templates (<code>string-template</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/4f4a79ee9f7ddb70b34647b219d3edbbe218adb7"><code>4f4a79e</code></a>: Standardize import order (<code>import-ordering</code>)

<details>
<summary><a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/511be48811088be8fcdb496a385ec58879130151"><code>511be48</code></a>: Make wildcard imports explicit (<code>no-wildcard-imports</code>)</summary>

* @matkoniecz [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015620896): is it going to be undone by Android Studio? Sometimes its "autofix missing import" is doing this.
* @FloEdelmann [wrote](https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100): Actually, I fixed it manually by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>o</kbd> (optimize imports) in Android Studio. There were only a few cases where that didn't work; there I removed one wildcard import at a time and used Android Studio's import suggestions. So it seems that Android Studio won't reintroduce wildcard imports by itself.

</details>

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/b622664109ce3cae090660891827efd0b2e96781"><code>b622664</code></a>: Remove unused imports (<code>no-unused-imports</code>)

<a href="https://github.com/streetcomplete/StreetComplete/pull/3649/commits/96dea182a2eba0a65329128e44e69723324d91fc"><code>96dea18</code></a>: Rename file to match class name (<code>filename</code>)

smichel17 commented:
> it seems that Android Studio won't reintroduce wildcard imports by itself.

There's a setting. I believe the default is that it switches to a wildcard if there's more than 3 imports from the same package. I always turn it to "never". Maybe that's already set in the configuration file now checked into the repo.

FloEdelmann commented:
> Maybe that's already set in the configuration file now checked into the repo.

Exactly. That's defined in https://github.com/streetcomplete/StreetComplete/blob/ffdac58b577d69b78b4f3f89c6d506e6385a2fc8/.idea/codeStyles/Project.xml#L9-L10

smichel17 commented:
I'm much in favor of `&&` prefixing. It's easier for me to follow and, in my opinion, makes it mistakes like https://github.com/streetcomplete/StreetComplete/commit/fb8fdd44e74a129d35d5174dbc2adcd0ba4146f7 more obvious (although the linter will not catch them on its own).

westnordost commented:
I've finally took the time to look at this.

> Question to @westnordost and others: Do you like the idea of an automated linter at all?
> 1. If yes, then we need to decide if we want to disable any rules. I'd say no, but let's hear your opinions. To help with the decision, please go through this PR's commits that start with Lint: and see what code changes the respective rule causes. Please also see this comment for discussions about specific rules.
> 2. The rules that we keep enabled should be fixed, and the resulting formatting should look good. That means, that I myself will go through the commits once more, and check the output thoroughly. Where needed, I'll add more manual lint changes that are in line with the code style.
> 3. We need to decide what to do with open PRs. The big list of changes will cause merge conflicts all over the place. Should we take the time once to fix the conflicts in all open PRs? I'd volunteer to do that where I have the privileges to do so.

So-so. The rules as summarized by you are okay. I'd not like to loose the ❔stuff though, I think it makes it more readable.

What I definitely don't want, and I think Mateusz mentioned it before, is some automatic lint check on GitHub, that prompts users to change their formatting like [seen here](https://github.com/FloEdelmann/StreetComplete/pull/508).
Instead, if this is possible, this should be done via Android Studio **while** typing the code. Best solution would be if there was a way to configure this similar to `.editorconfig` so that everyone who checks out this project and uses Android Studio will get the same code style applied automatically. Is this possible?
If this is not possible, then running this lint every now and then manually (via Android studio, or gradle, or Github Actions) would be fine for me. After a first run of this linter, I don't expect much inconsistencies growing back into the codebase again.

westnordost commented on app/src/androidTest/java/de/westnordost/streetcomplete/quests/existence/CheckExistenceTitleTest.kt:
```diff
@@ -33,56 +33,68 @@ class CheckExistenceTitleTest {
-            "operator" to "Deutsche Post AG",
-            "ref" to "Hauptsmoorstr. 101, 96052 Bamberg",
-        ))
+        val postBox = getQuestTitle(
+            mapOf(
+                "amenity" to "post_box",
+                "brand" to "Deutsche Post",
+                "operator" to "Deutsche Post AG",
+                "ref" to "Hauptsmoorstr. 101, 96052 Bamberg",
+            )
+        )
```
Which rule is this? This was deliberate, to keep the number of indentations as low as possible because I think it does improve redability.

westnordost commented on app/src/androidTest/java/de/westnordost/streetcomplete/data/osm/mapdata/RelationDaoTest.kt:
```diff
@@ -79,44 +85,56 @@ class RelationDaoTest : ApplicationDbTestCase() {
-            RelationMember(WAY, 4, "two"),
-        ))
-        dao.putAll(listOf(e1,e2,e3))
+        val e2 = rel(
+            2,
+            members = listOf(
+                RelationMember(NODE, 0, "bla"),
+                RelationMember(WAY, 1, "blub"),
+            )
+        )
+        val e3 = rel(
+            3,
+            members = listOf(
+                RelationMember(RELATION, 3, "one"),
+                RelationMember(WAY, 4, "two"),
+            )
+        )
```
Which rule is this? I find the way it was before more readable.

FloEdelmann replied:
Again, the `indent` rule ([`9c2f170`](https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9c2f170cfb7c138d01501e0c7ccbecc54f41b511#diff-1da3edd1bb0a07d6737c2859252bff1d67eb85cdbccc9219024ac5fdb485e6b8)).

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/AndroidDatabase.kt:
```diff
@@ -3,11 +3,33 @@ package de.westnordost.streetcomplete.data
 import android.database.sqlite.SQLiteStatement
-import de.westnordost.streetcomplete.data.ConflictAlgorithm.*
-import de.westnordost.streetcomplete.ktx.*
+import de.westnordost.streetcomplete.data.ConflictAlgorithm.ABORT
+import de.westnordost.streetcomplete.data.ConflictAlgorithm.FAIL
+import de.westnordost.streetcomplete.data.ConflictAlgorithm.IGNORE
+import de.westnordost.streetcomplete.data.ConflictAlgorithm.REPLACE
+import de.westnordost.streetcomplete.data.ConflictAlgorithm.ROLLBACK
+import de.westnordost.streetcomplete.ktx.getBlob
+import de.westnordost.streetcomplete.ktx.getBlobOrNull
+import de.westnordost.streetcomplete.ktx.getDouble
+import de.westnordost.streetcomplete.ktx.getDoubleOrNull
+import de.westnordost.streetcomplete.ktx.getFloat
+import de.westnordost.streetcomplete.ktx.getFloatOrNull
+import de.westnordost.streetcomplete.ktx.getInt
+import de.westnordost.streetcomplete.ktx.getIntOrNull
+import de.westnordost.streetcomplete.ktx.getLong
+import de.westnordost.streetcomplete.ktx.getLongOrNull
+import de.westnordost.streetcomplete.ktx.getShort
+import de.westnordost.streetcomplete.ktx.getShortOrNull
+import de.westnordost.streetcomplete.ktx.getString
+import de.westnordost.streetcomplete.ktx.getStringOrNull
```
Why are wildcard imports considered to be bad?

smichel17 replied:
Wildcard imports make it harder to browse the source code as plain text. Explicit imports means I can search in the file to find where something is imported from.

Why are explicit imports bad? That's a rhetorical question; we've had this conversation before, so I know the answer. The point of the question is to show that "you should use better tooling" works in both directions.

A smart IDE can auto-import and collapse the imports, making them effectively equivalent to wildcards, for people using the IDE. People who opt for simple tooling do not have this option, so we should style for them by default.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/edithistory/Edit.kt:
```diff
@@ -14,7 +14,7 @@ interface Edit {
-data class NoteEditKey(val id: Long): EditKey()
-data class OsmQuestHiddenKey(val osmQuestKey: OsmQuestKey): EditKey()
-data class OsmNoteQuestHiddenKey(val osmNoteQuestKey: OsmNoteQuestKey): EditKey()
+data class ElementEditKey(val id: Long) : EditKey()
+data class NoteEditKey(val id: Long) : EditKey()
+data class OsmQuestHiddenKey(val osmQuestKey: OsmQuestKey) : EditKey()
+data class OsmNoteQuestHiddenKey(val osmNoteQuestKey: OsmNoteQuestKey) : EditKey()
```
This seems inconsistent to me. So here, there should be a `:`. But at 

```
try { ... } catch (e: Exception) { ... }
```
not?

And also not at 
```
fun myFun(): Boolean {
  return true
}
```
([Which rule is this? This was deliberate, to keep the ](https://github.com/streetcomplete/StreetComplete/pull/3649/files#diff-8643389ef83ba98cd5bac5c2b9cfaa0e89f78501d979a3dc97ca44ebcb737d82))

smichel17 replied:
From https://kotlinlang.org/docs/coding-conventions.html#colon

> Put a space before : in the following cases:
> - when it's used to separate a type and a supertype
> - when delegating to a superclass constructor or a different constructor of the same class
> - after the object keyword
>
> Don't put a space before : when it separates a declaration and its type.

smichel17 replied:
I have no opinion on this, just sharing where it came from.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/elementfilter/ElementFilterExpression.kt:
```diff
@@ -19,9 +33,9 @@ class ElementFilterExpression(
-        includesElementType(element.type)
-        && (element.tags.isNotEmpty() || mayEvaluateToTrueWithNoTags)
-        && (elementExprRoot?.matches(element) ?: true)
+        includesElementType(element.type) &&
+            (element.tags.isNotEmpty() || mayEvaluateToTrueWithNoTags) &&
+            (elementExprRoot?.matches(element) ?: true)
```
Why would that be better than previously?

FloEdelmann replied:
It isn't; we already came to that conclusion in the summary. I didn't want to start with reverting specific commits just yet though.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/data/elementfilter/OverpassQueryCreator.kt:
```diff
@@ -38,14 +41,16 @@ class OverpassQueryCreator(
-            WAYS -> "way"
-            RELATIONS -> "rel"
-        } }
+        containsAll(listOf(NODES, WAYS, RELATIONS)) -> listOf("nwr")
+        containsAll(listOf(NODES, WAYS)) -> listOf("nw")
+        containsAll(listOf(WAYS, RELATIONS)) -> listOf("wr")
```
See my first comment, I think it is not as well readable as before.

smichel17 replied:
Mentioned in https://github.com/streetcomplete/StreetComplete/pull/3649#issuecomment-1015684100 but it didn't make its way into the summary. Direct answer: https://github.com/streetcomplete/StreetComplete/pull/3582#discussion_r766667403 (in the above comment).

westnordost commented on app/src/test/java/de/westnordost/streetcomplete/quests/AddCyclewayTest.kt:
```diff
@@ -65,16 +96,24 @@ class AddCyclewayTest {
+                )
+            )
+        )
+        val p1 = p(0.0, 0.0)
```
This is a good example for why I think the style before was better = easier to read.

Maybe this is a matter of taste, but I got a problem with this deeper level on indentation and all those free floating `)` everywhere

FloEdelmann replied:
Again, the `indent` rule ([`9c2f170`](https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9c2f170cfb7c138d01501e0c7ccbecc54f41b511#diff-33398a17f9b4bdf60a442b7ceee4338601d22794b2884515f80504bf63cb216c)).

westnordost commented on app/src/test/java/de/westnordost/streetcomplete/quests/AddPostboxCollectionTimesTest.kt:
```diff
@@ -22,25 +24,37 @@ class AddPostboxCollectionTimesTest {
+                        },
+                    )
+                )
+            ),
```
This is another. TBH, I really don't like a linter that would force when to make a newline and when not to. This decision should be up to the human, because in the end, the human also needs to read this. Linters should really limit themselves to small stuff like removing dangling whitespaces, wrong indentation and stuff like this. Not stuff that may be a matter of taste.

FloEdelmann replied:
And again, the `indent` rule ([`9c2f170`](https://github.com/streetcomplete/StreetComplete/pull/3649/commits/9c2f170cfb7c138d01501e0c7ccbecc54f41b511#diff-906eaa22cdf62d42a511403ae9c99aa0e6cf1c69048042d59138f17ffb239e04)).

FloEdelmann commented:
Thanks for the review @westnordost!

Regarding the rules: The ones you don't like are `indent` (which hasn't been raised before by @matkoniecz and @smichel17), the :grey_question: ones (`paren-spacing` and `no-multi-spaces`) and the :no_entry: one (`chain-wrapping`). Let's disable them all then. I'll update the summary with your comments.

> What I definitely don't want, and I think Mateusz mentioned it before, is some automatic lint check on GitHub, that prompts users to change their formatting when they create a PR, as [seen here](https://github.com/FloEdelmann/StreetComplete/pull/508).

Why not? Do you think it's intimidating for new contributors? Or that it's a waste of time for them?

I'd argue that it actually helps save time for both contributors and reviewers: When commits with lint issues are pushed to a PR, the contributor won't know at first. Then once a reviewer looks over the PR and notices a lint error, they have to make a decision whether to ask the contributor to fix it, or do it themselves now or later, or just ignore it.

Only in the first case will the contributor learn *that* they should have formatted something in specific way; and they will only learn "*why*" if the reviewer has put effort into explaining that. An automatic lint check can do that immediately without the contributor having to wait for a review. And we can [customize that message](https://github.com/streetcomplete/StreetComplete/pull/3649/commits/ffdac58b577d69b78b4f3f89c6d506e6385a2fc8) to make it less intimidating and more helpful. Also note that a failing check does *not* prevent merging (although you could require certain checks to pass in the repo settings).

So in the best case, this exact lint error is avoided right from the start the next time. In the worst case, the contributor is so intimidated that they refrain from ever contributing again. I find that quite unlikely.

> After a first run of this linter, I don't expect much inconsistencies growing back into the codebase again.

If you expect most PRs to be compliant, why are you against an automatic PR lint check then?

> Instead, if this is possible, this should be done via Android Studio **while** typing the code.

I agree that this would be the best case. Unfortunately, while there are Gradle plugins, I couldn't get them to run automatically during coding or building. They seem to require manual configuration for each user.
However, Android Studio's code style settings are now checked into git, so just formatting the code inside Android Studio now fixes most lint issues.

westnordost commented:
> >What I definitely don't want, and I think Mateusz mentioned it before, is some automatic lint check on GitHub, that prompts users to change their formatting when they create a PR, as seen here.

> Why not? Do you think it's intimidating for new contributors? Or that it's a waste of time for them?

Yes, both. And a waste of time for the reviewers to point to. The linter is not even about catching possible issues, but about code style (spaces after commas and so forth), it is _**really**_ unimportant.

> If you expect most PRs to be compliant, why are you against an automatic PR lint check then?

I expect them to be generally compliant with the code style because people contributing tend to copy the style the find.

> However, Android Studio's code style settings are now checked into git, so just formatting the code inside Android Studio now fixes most lint issues.

Well, that sounds good. Let's do that then. Is this already on master?

FloEdelmann commented:
I'll close this PR now, and create separate a PR that fixes the accepted lint rules.
